File: acuity/libraries/cpp/src/AcuityVision.cpp
// Acuity Vision — C++ client implementation.
//
// Reads the NetworkTables 4 schema documented in
// acuity/docs/nt4-schema.md. Same connection / staleness model as
// the Java + Python bindings: every getter checks heartbeat
// freshness and returns nullopt when the device has dropped.

#include "acuity/AcuityVision.h"

#include <chrono>
#include <cstdlib>
#include <cstring>
#include <mutex>
#include <string>
#include <utility>
#include <vector>

#include <networktables/DoubleTopic.h>
#include <networktables/IntegerTopic.h>
#include <networktables/NetworkTable.h>
#include <networktables/NetworkTableInstance.h>
#include <networktables/StringTopic.h>

namespace acuity {

// Heartbeat older than this means "the device is gone." See the Java
// binding for the rationale on the exact value.
static constexpr double kStaleAfterSeconds = 0.25;

// ---------------------------------------------------------------------
// Internal state
// ---------------------------------------------------------------------

namespace {

class AcuityImpl {
 public:
  AcuityImpl() {
    auto inst    = nt::NetworkTableInstance::GetDefault();
    m_root       = inst.GetTable("acuity");
    auto tags    = m_root->GetSubTable("tags");
    auto bestTag = tags->GetSubTable("best");
    auto bestObj = m_root->GetSubTable("objects")->GetSubTable("best");
    auto health  = m_root->GetSubTable("health");

    m_heartbeat        = m_root->GetIntegerTopic("heartbeat").Subscribe(0);
    m_bestId           = bestTag->GetIntegerTopic("id").Subscribe(-1);
    m_bestDistance     = bestTag->GetDoubleTopic("distance_m").Subscribe(0.0);
    m_bestYaw          = bestTag->GetDoubleTopic("yaw_deg").Subscribe(0.0);
    m_bestPitch        = bestTag->GetDoubleTopic("pitch_deg").Subscribe(0.0);
    m_bestTx           = bestTag->GetDoubleTopic("tx").Subscribe(0.0);
    m_bestTy           = bestTag->GetDoubleTopic("ty").Subscribe(0.0);
    m_bestArea         = bestTag->GetDoubleTopic("area").Subscribe(0.0);
    m_bestTimestamp    = bestTag->GetDoubleTopic("timestamp").Subscribe(0.0);
    m_bestDecisionMargin = bestTag->GetDoubleTopic("decision_margin").Subscribe(0.0);
    m_allTagsJson      = tags->GetStringTopic("all").Subscribe("[]");

    m_bestObjClass = bestObj->GetStringTopic("class").Subscribe("");
    m_bestObjConf  = bestObj->GetDoubleTopic("conf").Subscribe(0.0);
    m_bestObjTx    = bestObj->GetDoubleTopic("tx").Subscribe(0.0);
    m_bestObjTy    = bestObj->GetDoubleTopic("ty").Subscribe(0.0);

    m_cpuPct  = health->GetDoubleTopic("cpu_pct").Subscribe(0.0);
    m_tempC   = health->GetDoubleTopic("temp_c").Subscribe(0.0);
    m_uptimeS = health->GetIntegerTopic("uptime_s").Subscribe(0);
  }

  bool IsConnected() const {
    auto atomic = m_heartbeat.GetAtomic();
    if (atomic.time == 0) return false;
    double nowSec  = static_cast<double>(
        std::chrono::duration_cast<std::chrono::microseconds>(
            std::chrono::system_clock::now().time_since_epoch()).count()) / 1e6;
    double lastSec = static_cast<double>(atomic.time) / 1e6;
    return (nowSec - lastSec) < kStaleAfterSeconds;
  }

  std::optional<TagDetection> GetBestTag() const {
    if (!IsConnected()) return std::nullopt;
    int id = static_cast<int>(m_bestId.Get());
    if (id < 0) return std::nullopt;
    return TagDetection{
        id,
        m_bestDistance.Get(),
        m_bestYaw.Get(),
        m_bestPitch.Get(),
        m_bestTx.Get(),
        m_bestTy.Get(),
        m_bestArea.Get(),
        m_bestTimestamp.Get(),
        m_bestDecisionMargin.Get(),
    };
  }

  std::vector<TagDetection> GetAllTags() const {
    if (!IsConnected()) return {};
    std::string json = m_allTagsJson.Get();
    return ParseTagJsonArray(json);
  }

  std::optional<ObjectDetection> GetBestObject() const {
    if (!IsConnected()) return std::nullopt;
    std::string klass = m_bestObjClass.Get();
    if (klass.empty()) return std::nullopt;
    return ObjectDetection{std::move(klass),
                           m_bestObjConf.Get(),
                           m_bestObjTx.Get(),
                           m_bestObjTy.Get()};
  }

  DeviceHealth GetHealth() const {
    return DeviceHealth{
        m_cpuPct.Get(),
        m_tempC.Get(),
        static_cast<int64_t>(m_uptimeS.Get()),
    };
  }

 private:
  // Hand-rolled JSON parser — we only need to extract a flat list of
  // {"id":..., "distance_m":..., ...} objects, and pulling in nlohmann
  // / RapidJSON as a vendordep transitive dep isn't worth it.
  static std::vector<TagDetection> ParseTagJsonArray(const std::string& s) {
    std::vector<TagDetection> out;
    size_t i = 0, n = s.size();
    while (i < n && s[i] != '[') i++;
    if (i == n) return out;
    i++;
    while (i < n) {
      while (i < n && (std::isspace(static_cast<unsigned char>(s[i])) || s[i] == ',')) i++;
      if (i >= n || s[i] == ']') break;
      if (s[i] != '{') break;
      size_t objStart = i;
      int depth = 0;
      while (i < n) {
        char c = s[i];
        if (c == '{') depth++;
        else if (c == '}') { depth--; if (depth == 0) { i++; break; } }
        else if (c == '"') {
          i++;
          while (i < n && s[i] != '"') {
            if (s[i] == '\\' && i + 1 < n) i++;
            i++;
          }
        }
        i++;
      }
      out.push_back(TagFromObject(s.substr(objStart, i - objStart)));
    }
    return out;
  }

  static TagDetection TagFromObject(const std::string& obj) {
    return TagDetection{
        static_cast<int>(ExtractDouble(obj, "id", -1)),
        ExtractDouble(obj, "distance_m", 0),
        ExtractDouble(obj, "yaw_deg", 0),
        ExtractDouble(obj, "pitch_deg", 0),
        ExtractDouble(obj, "tx", 0),
        ExtractDouble(obj, "ty", 0),
        ExtractDouble(obj, "area", 0),
        ExtractDouble(obj, "timestamp", 0),
        ExtractDouble(obj, "decision_margin", 0),
    };
  }

  static double ExtractDouble(const std::string& obj, const std::string& key, double def) {
    std::string needle = "\"" + key + "\"";
    auto k = obj.find(needle);
    if (k == std::string::npos) return def;
    auto colon = obj.find(':', k + needle.size());
    if (colon == std::string::npos) return def;
    auto j = colon + 1;
    while (j < obj.size() && std::isspace(static_cast<unsigned char>(obj[j]))) j++;
    auto start = j;
    while (j < obj.size()) {
      char c = obj[j];
      if (c == ',' || c == '}' || std::isspace(static_cast<unsigned char>(c))) break;
      j++;
    }
    try {
      return std::stod(obj.substr(start, j - start));
    } catch (...) {
      return def;
    }
  }

  std::shared_ptr<nt::NetworkTable> m_root;

  nt::IntegerSubscriber m_heartbeat;
  nt::IntegerSubscriber m_bestId;
  nt::DoubleSubscriber  m_bestDistance;
  nt::DoubleSubscriber  m_bestYaw;
  nt::DoubleSubscriber  m_bestPitch;
  nt::DoubleSubscriber  m_bestTx;
  nt::DoubleSubscriber  m_bestTy;
  nt::DoubleSubscriber  m_bestArea;
  nt::DoubleSubscriber  m_bestTimestamp;
  nt::DoubleSubscriber  m_bestDecisionMargin;
  nt::StringSubscriber  m_allTagsJson;

  nt::StringSubscriber m_bestObjClass;
  nt::DoubleSubscriber m_bestObjConf;
  nt::DoubleSubscriber m_bestObjTx;
  nt::DoubleSubscriber m_bestObjTy;

  nt::DoubleSubscriber  m_cpuPct;
  nt::DoubleSubscriber  m_tempC;
  nt::IntegerSubscriber m_uptimeS;
};

AcuityImpl& Impl() {
  // Lazy singleton — same lifetime as the parent process.
  static AcuityImpl impl;
  return impl;
}

}  // namespace

// ---------------------------------------------------------------------
// Public API — delegates to the singleton impl.
// ---------------------------------------------------------------------

AcuityVision::AcuityVision() = default;

AcuityVision& AcuityVision::GetInstance() {
  static AcuityVision instance;
  Impl();  // ensure constructed
  return instance;
}

std::optional<TagDetection> AcuityVision::GetBestTag() const {
  return Impl().GetBestTag();
}

std::vector<TagDetection> AcuityVision::GetAllTags() const {
  return Impl().GetAllTags();
}

std::optional<ObjectDetection> AcuityVision::GetBestObject() const {
  return Impl().GetBestObject();
}

DeviceHealth AcuityVision::GetHealth() const {
  return Impl().GetHealth();
}

bool AcuityVision::IsConnected() const {
  return Impl().IsConnected();
}

}  // namespace acuity

File: acuity/libraries/java/src/main/java/tech/acuity/AcuityVision.java
// Acuity Vision — Java client library.
//
// Thin wrapper around NetworkTables 4. Reads the schema documented in
// acuity/docs/nt4-schema.md and exposes a typed, idiomatic Java API
// so robot code never has to remember topic paths.
//
// Connection / staleness model:
//   * The library does not "connect" or "disconnect" — that's NT4's
//     job. We just observe whether the device's heartbeat topic has
//     been updated within `STALE_AFTER_SECONDS`.
//   * Every getter returns Optional.empty() (or a stale-flagged
//     DeviceHealth) when the device is gone, so robot code never sees
//     stale tag data treated as live.

package tech.acuity;

import edu.wpi.first.networktables.DoubleSubscriber;
import edu.wpi.first.networktables.IntegerSubscriber;
import edu.wpi.first.networktables.NetworkTable;
import edu.wpi.first.networktables.NetworkTableInstance;
import edu.wpi.first.networktables.PubSubOption;
import edu.wpi.first.networktables.StringSubscriber;
import edu.wpi.first.networktables.Topic;

import java.util.ArrayList;
import java.util.List;
import java.util.Optional;

/**
 * Singleton entry point for the Acuity vision coprocessor.
 *
 * <pre>
 * AcuityVision vision = AcuityVision.getInstance();
 * vision.getBestTag().ifPresent(tag -> {
 *   double yaw = tag.yawDeg;
 * });
 * </pre>
 */
public final class AcuityVision {

  /**
   * Heartbeat older than this means "the device is gone." 100ms is
   * one Acuity heartbeat interval at 10 Hz, plus a safety margin —
   * tight enough that robot code reacts fast when the device drops,
   * loose enough that a single skipped heartbeat doesn't flap.
   */
  private static final double STALE_AFTER_SECONDS = 0.25;

  private static AcuityVision instance;

  private final NetworkTable        root;
  private final NetworkTable        bestTag;
  private final NetworkTable        tags;
  private final NetworkTable        bestObj;
  private final NetworkTable        health;

  private final IntegerSubscriber   heartbeat;
  private final IntegerSubscriber   bestId;
  private final DoubleSubscriber    bestDistance;
  private final DoubleSubscriber    bestYaw;
  private final DoubleSubscriber    bestPitch;
  private final DoubleSubscriber    bestTx;
  private final DoubleSubscriber    bestTy;
  private final DoubleSubscriber    bestArea;
  private final DoubleSubscriber    bestTimestamp;
  private final DoubleSubscriber    bestDecisionMargin;
  private final StringSubscriber    allTagsJson;

  private final StringSubscriber    bestObjClass;
  private final DoubleSubscriber    bestObjConf;
  private final DoubleSubscriber    bestObjTx;
  private final DoubleSubscriber    bestObjTy;

  private final DoubleSubscriber    cpuPct;
  private final DoubleSubscriber    tempC;
  private final IntegerSubscriber   uptimeS;

  private AcuityVision() {
    NetworkTableInstance nt = NetworkTableInstance.getDefault();
    root    = nt.getTable("acuity");
    tags    = root.getSubTable("tags");
    bestTag = tags.getSubTable("best");
    bestObj = root.getSubTable("objects").getSubTable("best");
    health  = root.getSubTable("health");

    PubSubOption[] opts = { PubSubOption.keepDuplicates(false) };

    heartbeat          = root.getIntegerTopic("heartbeat").subscribe(0, opts);
    bestId             = bestTag.getIntegerTopic("id").subscribe(-1, opts);
    bestDistance       = bestTag.getDoubleTopic("distance_m").subscribe(0.0, opts);
    bestYaw            = bestTag.getDoubleTopic("yaw_deg").subscribe(0.0, opts);
    bestPitch          = bestTag.getDoubleTopic("pitch_deg").subscribe(0.0, opts);
    bestTx             = bestTag.getDoubleTopic("tx").subscribe(0.0, opts);
    bestTy             = bestTag.getDoubleTopic("ty").subscribe(0.0, opts);
    bestArea           = bestTag.getDoubleTopic("area").subscribe(0.0, opts);
    bestTimestamp      = bestTag.getDoubleTopic("timestamp").subscribe(0.0, opts);
    bestDecisionMargin = bestTag.getDoubleTopic("decision_margin").subscribe(0.0, opts);
    allTagsJson        = tags.getStringTopic("all").subscribe("[]", opts);

    bestObjClass = bestObj.getStringTopic("class").subscribe("", opts);
    bestObjConf  = bestObj.getDoubleTopic("conf").subscribe(0.0, opts);
    bestObjTx    = bestObj.getDoubleTopic("tx").subscribe(0.0, opts);
    bestObjTy    = bestObj.getDoubleTopic("ty").subscribe(0.0, opts);

    cpuPct  = health.getDoubleTopic("cpu_pct").subscribe(0.0, opts);
    tempC   = health.getDoubleTopic("temp_c").subscribe(0.0, opts);
    uptimeS = health.getIntegerTopic("uptime_s").subscribe(0, opts);
  }

  /** Returns the global Acuity client, creating it lazily. */
  public static synchronized AcuityVision getInstance() {
    if (instance == null) instance = new AcuityVision();
    return instance;
  }

  /**
   * Largest AprilTag in the camera's current frame, or empty if no
   * tag is in view (or the device is disconnected).
   */
  public Optional<TagDetection> getBestTag() {
    if (!isConnected()) return Optional.empty();
    int id = (int) bestId.get();
    if (id < 0) return Optional.empty();
    return Optional.of(new TagDetection(
        id,
        bestDistance.get(),
        bestYaw.get(),
        bestPitch.get(),
        bestTx.get(),
        bestTy.get(),
        bestArea.get(),
        bestTimestamp.get(),
        bestDecisionMargin.get()));
  }

  /**
   * Every AprilTag in the current frame. Empty list if none in view
   * or device is disconnected. Parsed once per call from the JSON
   * blob the device publishes — cheap, but call once per loop and
   * cache if you're iterating heavily.
   */
  public List<TagDetection> getAllTags() {
    if (!isConnected()) return List.of();
    String json = allTagsJson.get();
    if (json == null || json.isEmpty() || json.equals("[]")) return List.of();
    return TagDetection.parseJsonArray(json);
  }

  /** Highest-confidence detected object, if object detection is enabled. */
  public Optional<ObjectDetection> getBestObject() {
    if (!isConnected()) return Optional.empty();
    String klass = bestObjClass.get();
    if (klass == null || klass.isEmpty()) return Optional.empty();
    return Optional.of(new ObjectDetection(
        klass, bestObjConf.get(), bestObjTx.get(), bestObjTy.get()));
  }

  /** Current device health snapshot. Always returns; check {@link #isConnected()} for liveness. */
  public DeviceHealth getHealth() {
    return new DeviceHealth(cpuPct.get(), tempC.get(), uptimeS.get());
  }

  /** True if heartbeat has been seen within the staleness window. */
  public boolean isConnected() {
    long lastChangeMicros = heartbeat.getAtomic().timestamp;
    if (lastChangeMicros == 0) return false;  // never received
    double nowSec  = NetworkTableInstance.getDefault().getServerTimeOffset()
                       .map(off -> (System.currentTimeMillis() * 1000.0 + off) / 1e6)
                       .orElse(System.currentTimeMillis() / 1e3);
    double lastSec = lastChangeMicros / 1e6;
    return (nowSec - lastSec) < STALE_AFTER_SECONDS;
  }
}
